### PR TITLE
Improve fulfillment option at checkout

### LINKED ISSUE #613: No fulfillment option should be selected by default at checkout
When we checkout, we need to make fulfillment options selectable rather than it defaulting to a pre-selected option.

### PR DESCRIPTION
**Pull request by DeeTheDev**
Fixes #613 

This change infers that checkouts are created without a fulfillment option until they are `.completed` or an order is created with it.
- Create migration to make checkout `fulfillment_option_id` nullable
- Use fulfillment option from previous editable checkout for new checkouts if it exists
- Ensure that specs are working properly, specially orders commerce and checkout specs
- Update fixtures to represent this change

---

- Add new commerce `modify_fulfillment` endpoint to update checkout fulfillment option
- Test

---

Add validation to require checkout fulfillment option when completing a checkout

---

Since fulfillment option is now nullable at checkout, it's best to add form validation error handling
- Setup API to update checkout fulfillment option
- Use it in the checkout page to save the option to checkout
- Use `react-hook-form` for form validation
- Creating reusable `FormRadioInputs` component to handle multiple radio inputs with error handling from react-hook-form
- Use component in checkout page payment and fulfillment options inputs.
- Add and localize invalid radio/checkout input message

### Form validation error handling examples (En)
Screenshot 2024-04-03 at 1 00 51 PM
Screenshot 2024-04-03 at 1 01 25 PM

### Place order button is not disabled, we handle errors in the client/backend level too
Screenshot 2024-04-03 at 1 02 45 PM

### CODE REVIEW

File: lib/suma/api/commerce.rb [hunk 1]
```diff
@@ -145,11 +160,10 @@ def lookup_editable!
             checkout.payment_instrument = instrument if instrument
           end
 
-          if (fuloptid = params[:fulfillment_option_id])
-            fulopt = checkout.cart.offering.fulfillment_options_dataset[fuloptid]
-            merror!(403, "Fulfillment option not found", code: "resource_not_found") unless fulopt
-            checkout.fulfillment_option = fulopt
-          end
+          fuloptid = params[:fulfillment_option_id]
```
Comment (rgalanakis) [inline lib/suma/api/commerce.rb:163]: Why is this code so different from `modify_fulfillment`?

File: lib/suma/fixtures/checkouts.rb [hunk 2]
```diff
@@ -27,6 +25,8 @@ module Suma::Fixtures::Checkouts
 
   decorator :completed, presave: true do |t=Time.now|
```
Comment (rgalanakis) [inline lib/suma/fixtures/checkouts.rb:26]: Also add this:

```rb
decorator :with_fulfillment_option, presave: true do |o={}|
  o = Suma::Fixtures.offering_fulfillment_option(offering: self.cart.offering).create(o) unless o.is_a?(Suma::Offering::FulfillmentOption)
  self.fulfillment_option = o
end
```

And write a test for it so it's covered.

File: spec/suma/api/commerce_spec.rb [hunk 3]
```diff
@@ -151,13 +151,24 @@
...
-          fulfillment_option_id: fulfillment.id,
+          fulfillment_option_id: nil,
           available_fulfillment_options: contain_exactly(include(id: fulfillment.id)),
         )
       expect(other_member_checkout.refresh).to be_soft_deleted
       expect(completed_checkout.refresh).to_not be_soft_deleted
     end
 
+    it "starts a checkout with fulfillment option from a previously editable checkout" do
+      existing_editable_checkout = Suma::Fixtures.checkout(cart:, fulfillment_option: fulfillment).create
```
Comment (rgalanakis) [inline spec/suma/api/commerce_spec.rb:162]: Add a non-editable checkout too.

File: spec/suma/api/commerce_spec.rb [hunk 4]
```diff
@@ -272,10 +312,30 @@
       expect(checkout.refresh).to be_completed
     end
 
+    it "errors if fulfillment option id is missing" do
```
Comment (rgalanakis) [inline spec/suma/api/commerce_spec.rb:315]: Do not write tests for `requires` vs. `optional`, and other trivial settings in the `params` block.

File: lib/suma/api/commerce.rb (hunk 1 above)
Comment (DeeTheDev) [reply lib/suma/api/commerce.rb:163]: They need to be consistent thanks for catching this, fixing

**Pull request merged**